# Cross-Entropy

Cross-Entropy is a known method also in Statistics and Machine Learning. 
The thought process for the decision to use it goes back from the use of Probability.

(1) Probability -> We want a Maximum Likelihood 

Problem: Products are expensive

(2) Solution: Use sums via Log Function since log(ab) = log a + log b
Problem: log gave out negative numbers

Solution:Negate it

**Resulting Interpretation:**

connection between probability (max likelihood) and error (min error) which is Cross-Entropy
- The higher the cross entropy, the higher the error, the lower the probability (vice versa) 

Cross-Entropy formula:
$$CrossEntropy = -[ \sum_{}^{} y_{i} ln(P_{i}) + (1-y_{i}) ln(1-P_{i}) ]$$

In [3]:
import numpy as np

# Write a function that takes as input two lists Y, P,
# and returns the float corresponding to their cross-entropy.
def cross_entropy(Y, P):
    CE = np.sum([(-1*(Y[i]*np.log(P[i]) + (1-Y[i])*np.log(1-P[i]))) for i in range(0,len(Y))])
    return CE

ERROR! Session/line number was not unique in database. History logging moved to new session 59


In [4]:
cross_entropy(Y = [1,0,1,1], P = [0.4, 0.6, 0.1, 0.5])

4.828313737302301

Trying for Y=[1,0,1,1] and P=[0.4,0.6,0.1,0.5].

The correct answer is
4.8283137373
And your code returned
4.8283137373

Correct!

## Multi-Class Cross-Entropy

The idea is still the same and we can further extend the binary Cross Entropy as:

$$MultiClassCrossEntropy = -[  \sum_{i=1}^{n}\sum_{j=1}^{m} y_{ij} ln(P_{ij})]$$